### ISSUE TITLE
Possible to have the ability to turn on/off appended JSON when Oj::ParseError is raised

### ISSUE THREAD

**Issue by randyv12** [ohler55/oj#544]
Hello,

Would really appreciate it if I can get some help or direction on how to omit the JSON string at the end of the error message when an OJ::ParseError is raised in compat mode.

Currently when Oj.load is called in :compat mode and an Oj::ParseError is raised, we get an error with the helpful string displayed at the end of [parse.c:704]:

```ruby
2.5.1 :011 > Oj.load("{a", mode: :compat)
```
```
Oj::ParseError (unexpected character () at line 1, column 2 [parse.c:704] in '{a)
```
But in :rails mode we don't see the helpful message displayed in the ParseError object

```ruby
2.5.1 :010 > Oj.load("{a", mode: :rails)
```
```
Oj::ParseError (unexpected character () at line 1, column 2 [parse.c:704])
```

I was wondering if we can disable the appended ```in '{a)``` message with a default_options flag, in compat mode.

Currently, when ```Oj.optimize_rails()``` is used, the default mode turns into compat mode.  So it's difficult, (but possible) to omit the appended json string at the end of the error displayed.

Also, my ultimate goal is to have Oj parse request params in Rails without the appended JSON body in the error.message string.
We can see that the error.message is passed unto a new message with e.message here:
https://github.com/rails/rails/blob/4-2-stable/actionpack/lib/action_dispatch/middleware/params_parser.rb#L53

```ruby
ActiveSupport::JSON.decode("{a:1")
```
to output:
```
JSON::ParserError (unexpected character () at line 1, column 2 [parse.c:704]
```
instead of:
```
JSON::ParserError (unexpected character () at line 1, column 2 [parse.c:704] in '{a:1)
```

**Comment by ohler55** [comment]
I though the error messages had been normalized. I don't think we need an option but rather the ` in xxx` should just be removed.

**Comment by randyv12** [comment]
I think the original intent was to mimic the json gem.  And I think it's working as intended since it behaves like the json gem completely.

We see it being set here
https://github.com/ohler55/oj/blob/master/ext/oj/parse.c#L1039
and compat mode is set here during mimic json 
https://github.com/ohler55/oj/blob/master/ext/oj/mimic_json.c#L368

I agree though, that it should be removed.  The two  reason being that it's still debuggable but sometimes we run into the trouble that if the error is raised, integrated gems that do error reporting store these messages containing the json payload, and can sometimes be reflected back if rails is misconfigured to show full error reporting.

My guess is that, when we declare in an initializer 
```ruby
Oj.optimize_rails()
Oj.default_options =  {mode: :rails}
```
that setting the default mode won't be picked up on optimize rails because during runtime, the mode was already set when JSON.parse is mimicked

I think removing it makes the error message more simpler and improves the security of the usage of Oj.  

I really appreciate it if we can get this to work :D I wish I can help out in contributing code changes but I don't think I have enough expertise compiling and testing the Oj gem.

Thanks for the response, in the meantime using a monkey patch for ActiveSupport::JSON.decode to use Oj.load(json, quirks_mode: true, mode: :rails) should work

```
# rails 4 initializer
module ActiveSupport
  module JSON
    class << self
      def decode(json, options = {})
        if options.present?
          raise ArgumentError, "In Rails 4.1, ActiveSupport::JSON.decode no longer " \
            "accepts an options hash for MultiJSON. MultiJSON reached its end of life " \
            "and has been removed."
        end
        data = Oj.load(json, quirks_mode: true)
        if ActiveSupport.parse_json_times
          convert_dates_from(data)
        else
          data
        end
      end
    end
   end
 end
end
```

**Comment by ohler55** [comment]
Adding an option to make the JSON gem safer seems like the right option. Normally I would say being exactly like the JSON gem in mimic is best but in this case it really is a security issue. An error on a gigabyte file could easily cause unexpected problems in an application.

**Comment by ohler55** [comment]
You can take a look at the `buf-size` branch for the addition of the `safe` option. I have not updated the doc file yet but the code is in place as is the ruby docs.

**Comment by randyv12** [comment]
Thanks ! Tried it and it works.  Sorry for the late reply.
When tried the buf-size branch, I now get the parse.c error without the JSON appended at the end

**After:**
```
# rails 4.2.11 and ruby 2.5.1
ActionDispatch::ParamsParser::ParseError (unexpected character () at line 1, column 1 [parse.c:624]):
  actionpack (4.2.11.1) lib/action_dispatch/middleware/params_parser.rb:52:in `rescue in parse_formatted_parameters'
  actionpack (4.2.11.1) lib/action_dispatch/middleware/params_parser.rb:31:in `parse_formatted_parameters'
  actionpack (4.2.11.1) lib/action_dispatch/middleware/params_parser.rb:23:in `call'
```
**Before:**
```
ActionDispatch::ParamsParser::ParseError (unexpected character () at line 1, column 1 [parse.c:623] in '{):
  actionpack (4.2.11.1) lib/action_dispatch/middleware/params_parser.rb:52:in `rescue in parse_formatted_parameters'
  actionpack (4.2.11.1) lib/action_dispatch/middleware/params_parser.rb:31:in `parse_formatted_parameters'
  actionpack (4.2.11.1) lib/action_dispatch/middleware/params_parser.rb:23:in `call'
```

Thank you so much, really appreciate the help!
For anyone that is interested the oj.rb initializer for this looks like
```ruby
# config/initializers/oj.rb
Oj.default_options = {mode: :compat, safe: true}
Oj.optimize_rails()
```

**Comment by ohler55** [comment]
Great, waiting for feedback on one other issue and then I'll release.

**Comment by ohler55** [comment]
Released v3.8.0.

**Comment by randyv12** [comment]
Awesome! Thank you, we can close this now
